packages/vite/src/node/plugins/asset.ts

```diff
@@ -171,11 +171,13 @@ async function fileToBuiltUrl(
   let url
   if (
     config.build.lib ||
-    (!file.endsWith('.svg') &&
-      content.length < Number(config.build.assetsInlineLimit))
+    Buffer.byteLength(content) < config.build.assetsInlineLimit
   ) {
-    // base64 inlined as a string
-    url = `data:${mime.getType(file)};base64,${content.toString('base64')}`
+    // svgs can be inlined without base64
+    url = file.endsWith('.svg')
+      ? `data:image/svg+xml;utf8,${content}`
```

Comment:
This won't escape unsafe characters. Might be worth taking a similar approach to `svg-url-loader`? See [Explanation](https://github.com/bhovhannes/svg-url-loader#svg-url-loader), [Usage](https://github.com/bhovhannes/svg-url-loader/blob/b5ffa239effd8647cf136c78b27463906b134306/src/loader.js#L35) and [Implementation](https://github.com/bhovhannes/svg-url-loader/blob/b5ffa239effd8647cf136c78b27463906b134306/src/loader.js#L59L63).

Reply:
The data uri has a mime type, so I'm skeptical that the "unsafe characters" specification applies here.

Reply:
This is a URI, so yes you need to escape unsafe characters or it won't work.
E.g. for `<path fill="#ffffff" />`, the `#` starts the URL fragment and everything breaks, so it needs to be escaped to `<path fill="%23ffffff" />`